geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBasePlatform.scala

```diff
@@ -0,0 +1,87 @@
+import org.opengis.feature.simple.SimpleFeature
+
+trait HBasePlatform extends HBaseFeatureIndex {
+
+  override protected def buildPlatformScanPlan(ds: HBaseDataStore,
+                                               filter: HBaseFilterStrategyType,
+                                               originalRanges: Seq[Query],
+                                               table: TableName,
+                                               hbaseFilters: Seq[HFilter],
+                                               toFeatures: (Iterator[Result]) => Iterator[SimpleFeature]): HBaseQueryPlan = {
+    // check if these Scans or Gets
+    // Only in the case of 'ID IN ()' queries will this be Gets
+    val scans = originalRanges.head match {
+      case t: Get  => configureGet(originalRanges, hbaseFilters)
+      case t: Scan => configureMultiRowRangeFilter(ds, originalRanges, hbaseFilters)
+    }
+
+    ScanPlan(filter, table, scans, toFeatures)
+  }
+
+  private def configureGet(originalRanges: Seq[Query], hbaseFilters: Seq[HFilter]): Seq[Scan] = {
+    val filterList = new FilterList(hbaseFilters: _*)
+    // convert Gets to Scans for Spark SQL compatibility
+    originalRanges.map { r =>
+      val g = r.asInstanceOf[Get]
+      val start = g.getRow
+      val end = IndexAdapter.rowFollowingRow(start)
+      new Scan(g).setStartRow(start).setStopRow(end).setFilter(filterList).setSmall(true)
+    }
+  }
+
+  private def configureMultiRowRangeFilter(ds: HBaseDataStore, originalRanges: Seq[Query], hbaseFilters: Seq[HFilter]) = {
+    import scala.collection.JavaConversions._
+    val rowRanges = Lists.newArrayList[RowRange]()
+    originalRanges.foreach { r =>
+      rowRanges.add(new RowRange(r.asInstanceOf[Scan].getStartRow, true, r.asInstanceOf[Scan].getStopRow, false))
+    }
+    val sortedRowRanges = MultiRowRangeFilter.sortAndMerge(rowRanges)
+    val numRanges = sortedRowRanges.length
+    val numThreads = ds.config.queryThreads
+    // TODO: parameterize this?
+    val rangesPerThread = math.min(ds.config.maxRangesPerExtendedScan, math.max(1, math.ceil(numRanges / numThreads * 2).toInt))
+    // TODO: align partitions with region boundaries
+    val groupedRanges = Lists.partition(sortedRowRanges, rangesPerThread)
+
+    // group scans into batches to achieve some client side parallelism
+    val groupedScans = groupedRanges.map { localRanges =>
+      // TODO: FIX
+      // currently, this constructor will call sortAndMerge a second time
+      // this is unnecessary as we have already sorted and merged above
+      val mrrf = new MultiRowRangeFilter(localRanges)
+      val filterList = new FilterList(FilterList.Operator.MUST_PASS_ALL, mrrf)
+      hbaseFilters.foreach { f => filterList.addFilter(f) }
+
+      val s = new Scan()
+      s.setStartRow(localRanges.head.getStartRow)
+      s.setStopRow(localRanges.get(localRanges.length - 1).getStopRow)
+      s.setFilter(filterList)
+      // TODO: parameterize cache size
```

Comment:
Ok... since this PR is likely to get rebased, at that point, these TODO: should be turned into tickets.;)